### PR TITLE
Wip app refactor

### PR DESCRIPTION
**Pull request by ywalakamar**
**What does this PR do?**
- refactors the app execution workflow
- adds redis database for local caching
- creates the user endpoint

**How can this be reproduced?**
- kindly refer to the project docs

**Known issues**
- user password is not yet encrypted
- CI tests with CircleCi fails(unprovisioned)

**Screenshots**

insomnia

tests

### CODE REVIEW

File: ormconfig.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[
```
Comment (fatmali) [inline ormconfig.json]: Why did you delete the `ormconfig.js` file? The reason we were using it was to be able to pass credentials as environment variables in for production credentials, something we can't do with a `json` file. I suggest you declare your test object in the exported object of `ormconfig.js` file. Another important thing to note is credentials, even if they are test credentials are *never* committed. That's why we were using the `dotenv-safe` library to pass in the credentials as `env` variables. This way, it's not only secure, but offers the team flexibility in setting up their dev environments (e.g not everyone will have `admin` and `admin123` as username credentials).

File: ormconfig.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      "duration": 30000,
+      "options": {
+        "host": "localhost",
+        "port": 6379
+      }
+    },
+    "entities": ["dist/entity/*.js"],
+    "migrations": ["dist/migration/*.js"],
+    "cli": {
+      "entitiesDir": "src/entity",
```
Comment (fatmali) [inline ormconfig.json]: I suggest we stick with the `models` dir as the `entities` dir for now, we will do the refactor in a separate PR.

File: package.json [hunk 3]
```diff
@@ -17,15 +36,16 @@
...
   "bugs": {
-    "url": ""
+    "url": "https://github.com/DevCEldoret/devceldoret-backend/issues"
   },
   "homepage": "",
   "dependencies": {
     "body-parser": "^1.19.0",
     "cors": "^2.8.5",
-    "dotenv-safe": "^8.2.0",
+    "dotenv": "^8.2.0",
```
Comment (fatmali) [inline package.json]: The reason we were using `dotenv-safe` instead of `dotenv` is to be able to have a `.env.example` file that will guide contributors on what keys are required in the `.env` file

File: src/Utils/dbConnection.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { getConnectionOptions, createConnection } from "typeorm";
+import { NODE_ENV } from "./config";
+export const createConn = async () => {
+  try {
+    const connectionOptions = await getConnectionOptions(NODE_ENV);
+    console.log(connectionOptions);
```
Comment (fatmali) [inline src/Utils/dbConnection.ts]: remove this `console.log()`

File: src/Utils/dbConnection.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { getConnectionOptions, createConnection } from "typeorm";
+import { NODE_ENV } from "./config";
```
Comment (fatmali) [inline src/Utils/dbConnection.ts]: add `NODE_ENV` to the `.env.example` file

File: src/Utils/dbConnection.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { getConnectionOptions, createConnection } from "typeorm";
```
Comment (fatmali) [inline src/Utils/dbConnection.ts]: In my latest PR, to solve test issues. connection creation is in the `server.ts` file

File: src/Utils/error-handler.ts [hunk 7]
```diff
@@ -1,10 +1,9 @@
...
-      super(message);
-      this.status = status;
-    }
-  }
-  
-  export default ErrorHandler;
-  
\ No newline at end of file
+import { Response, NextFunction, Request } from "express";
+
```
Comment (fatmali) [inline src/Utils/error-handler.ts]: There is a PR also addressing this

File: src/controllers/event.ts [hunk 8]
```diff
@@ -1,89 +1,65 @@
...
 
 export async function getEvents(
   _req: Request,
   res: Response,
   next: NextFunction
 ) {
-  const conn = await connection;
-  const events = await conn.manager.find(Event);
-  res.status(200).send(events);
+  try {
```
Comment (fatmali) [inline src/controllers/event.ts]: This is also handled in another PR #32

File: src/controllers/location.ts [hunk 9]
```diff
@@ -1,45 +1,60 @@
...
-import connection from "../connection";
-import { Request, Response, NextFunction } from 'express';
+import { Request, Response, NextFunction } from "express";
+import { getRepository } from "typeorm";
 
-export async function getLocations(_req: Request, res: Response, next: NextFunction) {
-  const conn = await connection;
-  const locations = await conn.manager.find(Location);
-  res.send(locations);
+export async function getLocations(
```
Comment (fatmali) [inline src/controllers/location.ts]: addressed in PR #32

File: src/Utils/apiResponses.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        .send({ status: statusCode, message: message });
+    } catch (error) {
+      return response
+        .status(statusCode)
+        .send({ status: statusCode, message: error });
+    }
+  }
+}
+
+const prompt = new Responses();
```
Comment (fatmali) [inline src/Utils/apiResponses.ts]: what's the point of creating the `Responses` class if we are using the method statically? Why not just export the methods directly as functions?

File: src/Utils/apiResponses.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { Response } from "express";
+
+class Responses {
+  successMessage(response: Response, statusCode: number, message: String) {
```
Comment (fatmali) [inline src/Utils/apiResponses.ts]: good job on this helper functions 👏

File: src/app.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import * as express from "express";
```
Comment (fatmali) [inline src/app.ts]: This is a big change and I'd suggest it's addressed in a completely separate PR and branch to avoid conflicts with other PR's.

File: src/controllers/userController.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
+import { Request, Response, NextFunction } from "express";
+import { getRepository } from "typeorm";
+import { conn } from "../Utils/dbConnection";
+import { User } from "../entity/user";
+import prompt from "../Utils/apiResponses";
```
Comment (fatmali) [inline src/controllers/userController.ts]: If you were exporting the methods directly and not using a class, you would do this
```suggestion
import {successWithPayload, errorMessage} from "../Utils/apiResponses";
```

File: ormconfig.json (hunk 1 above)
Comment (ywalakamar) [reply ormconfig.json]: sounds good. WIP

File: src/Utils/dbConnection.ts (hunk 4 above)
Comment (ywalakamar) [reply src/Utils/dbConnection.ts]: wip

File: src/Utils/dbConnection.ts (hunk 5 above)
Comment (ywalakamar) [reply src/Utils/dbConnection.ts]: wip

File: package.json (hunk 3 above)
Comment (ywalakamar) [reply package.json]: sure

File: ormconfig.json (hunk 2 above)
Comment (ywalakamar) [reply ormconfig.json]: cool
